src/Compilers/CSharp/Portable/Binder/PatternExplainer.cs

```diff
@@ -0,0 +1,512 @@
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal static class PatternExplainer
+    {
+        /// <summary>
+        /// Find the shortest path from the root node to the node of interest.
+        /// </summary>
+        /// <param name="nodes">The set of nodes in topological order.</param>
+        /// <param name="node">The node of interest.</param>
+        /// <param name="nullPaths">Whether to permit following paths that test for null.</param>
+        /// <returns>The shortest path, excluding the node of interest.</returns>
+        private static ImmutableArray<BoundDecisionDagNode> ShortestPathToNode(
+            ImmutableArray<BoundDecisionDagNode> nodes,
+            BoundDecisionDagNode node,
+            bool nullPaths)
+        {
+            // compute the distance from each node to the endpoint.
+            var dist = PooledDictionary<BoundDecisionDagNode, int>.GetInstance();
+            int nodeCount = nodes.Length;
+            int distance(BoundDecisionDagNode x) => x != null && dist.TryGetValue(x, out int v) ? v : nodeCount + 2;
+            for (int i = nodeCount - 1; i >= 0; i--)
+            {
+                var n = nodes[i];
+                switch (n)
+                {
+                    case BoundEvaluationDecisionDagNode e:
+                        // evaluations are free; we are more interested in decision nodes
+                        dist.Add(e, distance(e.Next));
+                        break;
+                    case BoundTestDecisionDagNode { Test: BoundDagNonNullTest _ } t when !nullPaths:
+                        dist.Add(t, 1 + distance(t.WhenTrue));
+                        break;
+                    case BoundTestDecisionDagNode { Test: BoundDagExplicitNullTest _ } t when !nullPaths:
+                        dist.Add(t, 1 + distance(t.WhenFalse));
+                        break;
+                    case BoundTestDecisionDagNode t:
+                        dist.Add(t, 1 + Math.Min(distance(t.WhenTrue), distance(t.WhenFalse)));
+                        break;
+                    case BoundWhenDecisionDagNode w:
+                        dist.Add(w, 1 + Math.Min(distance(w.WhenTrue), distance(w.WhenFalse)));
+                        break;
+                    default:
+                        // treat the endpoint as distance 1.
+                        // treat other nodes as not on the path to the endpoint
+                        dist.Add(n, (n == node) ? 1 : nodeCount + 2);
+                        break;
+                }
+            }
+
+            // trace a path from the root node to the node of interest
+            var result = ArrayBuilder<BoundDecisionDagNode>.GetInstance(capacity: dist[nodes[0]]);
+            for (BoundDecisionDagNode n = nodes[0]; n != node;)
```

Comment:
>for (BoundDecisionDagNode n = nodes[0]; n != node;) [](start = 12, length = 51)

Is this duplicating the decisions from the first `for` loop about which path to take (`WhenTrue` or `WhenFalse`) at each node?

If so, consider storing the next node in the dictionary in the first `for` loop, so we can remove the duplication:
```
var dist = PooledDictionary<BoundDecisionDagNode,
    (int Distance, BoundDecisionDagNode Next)>.GetInstance();
```